### PR TITLE
[Speculative decoding] [Multi-Step] decouple should_modify_greedy_probs_inplace

### PR DESCRIPTION
**Pull request by SolitaryThinker**
Preparation PR for multi step. Not a blocker, but will make MS PR smaller and improve perf.

Decouple `should_modify_greedy_probs_inplace` from `include_gpu_probs_tensor` so that multi-step can set `include_gpu_probs_tensor` without also setting `should_modify_greedy_probs_inplace` and incurring the overhead of the probs modification (causes a GPU<>CPU sync). Not a blocker for multi-step, but does add ~1ms of GPU bubble between each step on A10G, will be a much bigger slow down on H100.

@cadedaniel This may be a perf bug for spec decode as well? ~~I also can't seem to get spec_decode tests to all consistently pass locally?~~

Torch Profile with multi-step and without decoupling (`should_modify_greedy_probs_inplace == True`)
image

cc @WoosukKwon @zhuohan123 @Yard1 @comaniac @rkooo567 

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by cadedaniel** [comment]
Will take a look tomorrow.

File: vllm/spec_decode/spec_decode_worker.py [hunk 1]
```diff
@@ -287,7 +287,10 @@ def _configure_model_sampler_for_spec_decode(self):
         """
         (self.scorer_worker.model_runner.model.sampler.include_gpu_probs_tensor
          ) = True
+        (self.scorer_worker.model_runner.model.sampler.
+         should_modify_greedy_probs_inplace) = True
```
Comment (cadedaniel) [inline vllm/spec_decode/spec_decode_worker.py:291]: Can we put this behind an interface so refactors to the worker / model / sampler do spread everywhere?

**Review by cadedaniel — COMMENTED** [review]
LGTM.
* Can we add a test for this? e.g. something that runs `include_gpu_probs_tensor` but without modified greedy probs, and we verify that the gpu probs tensor is there but the greedy probs are not modified.
* The modification of greedy probs should not incur a CPU sync 😄 . I'll take a look at that.

**Comment by cadedaniel** [comment]
By the way, what model is running in the profile ?

**Comment by SolitaryThinker** [comment]
This is llama 8B, but the profile is with multi-step, which has removed all the other sources of CPU syncs (CPU prepare_input, GPU<>CPU transfer for sampled token)

File: vllm/spec_decode/spec_decode_worker.py (hunk 1 above)
Comment (SolitaryThinker) [reply vllm/spec_decode/spec_decode_worker.py:299]: Not sure if there is a clean way to do this, as the line above sets `include_gpu_probs_tensor` before this PR also change the decoupled (this PR) `should_modify_greedy_probs_inplace`. Seems the scorer worker is passed around as `WorkerBase` so would need to add change that class which is not ideal. Am I missing something?

Comment (cadedaniel) [reply vllm/spec_decode/spec_decode_worker.py:299]: Oh right.. sorry I must have misread.

File: vllm/lora/layers.py [hunk 2]
```diff
@@ -1067,6 +1067,10 @@ def org_vocab_size(self):
     def include_gpu_probs_tensor(self):
         return self.base_layer.include_gpu_probs_tensor
 
+    @property
+    def should_modify_greedy_probs_inplace(self):
```
Comment (cadedaniel) [inline vllm/lora/layers.py:1071]: Confused why this is here, why does this layer need it?

Comment (SolitaryThinker) [reply vllm/lora/layers.py:1071]: I'm not sure, I just greped for all the places that `include_gpu_probs_tensor` was used and added the change there to keep semantics identical. See the lines above.

**Comment by SolitaryThinker** [comment]
Just added test for the inplace modification, let me know if there's anything else.

**Review by cadedaniel — APPROVED** [review]

**Pull request merged**